File: 200.岛屿数量.py
#
# @lc app=leetcode.cn id=200 lang=python3
#
# [200] 岛屿数量
#

# @lc code=start
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        # 上下左右
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def dfs(x, y):
            # 先判定越界，后判定是否是个还没有访问的陆地格子
            if x < 0 or y < 0 or x >= m or y >= n or grid[x][y] != "1":
                return
            grid[x][y] = "2"    # 剪枝，后续外层遍历不再访问
            for direction in directions:
                dfs(x + direction[0], y + direction[1])

        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":   # 只走还没有被访问的陆地格子
                    dfs(i, j)
                    res += 1            # dfs()内部只做访问和剪枝，外部才做结果记录
        return res
# @lc code=end

File: 200.岛屿数量.go
/*
 * @lc app=leetcode.cn id=200 lang=golang
 *
 * [200] 岛屿数量
 */

// @lc code=start
func numIslands(grid [][]byte) (res int) {
	// 将所有属于同一个岛屿的陆地格子全部更改为 2，后续遍历不再查看该格子
	m, n := len(grid), len(grid[0])
	directions := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	var dfs func(x, y int)
	dfs = func(x, y int) {
		// 先判定索引越界，后续遍历陆地格子
		if x < 0 || y < 0 || x >= m || y >= n || grid[x][y] != '1' {
			return
		}
		grid[x][y] = '2'
		for _, direction := range directions {
			dfs(x+direction[0], y+direction[1])
		}
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '1' { // 遍历每一个还没有被访问的陆地格子
				dfs(i, j)
				res++ // 只有 dfs() 外部可以增加结果，dfs()只做各个相邻陆地格子的访问和修改
			}
		}
	}
	return
}

// @lc code=end

